File: gams/apifiles/CSharp/Transport2/Transport2.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using GAMS;

namespace TransportSeq
{
    class Transport2
    {
        static void Main(string[] args)
        {
            GAMSWorkspace ws;
            if (Environment.GetCommandLineArgs().Length > 1)
                ws = new GAMSWorkspace(systemDirectory: Environment.GetCommandLineArgs()[1]);
            else
                ws = new GAMSWorkspace();
            // write an include file containing data with GAMS syntax
            using (StreamWriter writer = new StreamWriter(ws.WorkingDirectory + Path.DirectorySeparatorChar + "tdata.gms"))
            {
                writer.Write(GetDataText());
            }
            // run a job using an instance of GAMSOptions that defines the data include file
            using (GAMSOptions opt = ws.AddOptions())
            {
                GAMSJob t2 = ws.AddJobFromString(GetModelText());
                opt.Defines.Add("incname", "tdata");
                t2.Run(opt);
                foreach (GAMSVariableRecord rec in t2.OutDB.GetVariable("x"))
                    Console.WriteLine("x(" + rec.Keys[0] + "," + rec.Keys[1] + "): level=" + rec.Level + " marginal=" + rec.Marginal);
            }
        }

        static String GetDataText()
        {
            String data = @"
  Sets
       i   canning plants   / seattle, san-diego /
       j   markets          / new-york, chicago, topeka / ;

  Parameters

       a(i)  capacity of plant i in cases
         /    seattle     350
              san-diego   600  /

       b(j)  demand at market j in cases
         /    new-york    325
              chicago     300
              topeka      275  / ;

  Table d(i,j)  distance in thousands of miles
                    new-york       chicago      topeka
      seattle          2.5           1.7          1.8
      san-diego        2.5           1.8          1.4  ;

  Scalar f  freight in dollars per case per thousand miles  /90/ ;
";
            return data;
        }

static String GetModelText()
        {
            String model = @"
  Sets
       i   canning plants
       j   markets

  Parameters
       a(i)   capacity of plant i in cases
       b(j)   demand at market j in cases
       d(i,j) distance in thousands of miles
  Scalar f  freight in dollars per case per thousand miles;

$if not set incname $abort 'no include file name for data file provided'
$include %incname%

  Parameter c(i,j)  transport cost in thousands of dollars per case ;

            c(i,j) = f * d(i,j) / 1000 ;

  Variables
       x(i,j)  shipment quantities in cases
       z       total transportation costs in thousands of dollars ;

  Positive Variable x ;

  Equations
       cost        define objective function
       supply(i)   observe supply limit at plant i
       demand(j)   satisfy demand at market j ;

  cost ..        z  =e=  sum((i,j), c(i,j)*x(i,j)) ;

  supply(i) ..   sum(j, x(i,j))  =l=  a(i) ;

  demand(j) ..   sum(i, x(i,j))  =g=  b(j) ;

  Model transport /all/ ;

  Solve transport using lp minimizing z ;

  Display x.l, x.m ;
";

            return model;
        }

    }
}

File: gams/apifiles/C++/Transport2/Transport2.cpp
// Transport2.cpp : main project file.

#include "stdafx.h"
#include <iostream>

using namespace System;
using namespace System::IO;
using namespace GAMS;


static System::String^ GetDataText()
{
	System::String^ data ="Sets\n"+
		"    i   canning plants   / seattle, san-diego /\n"+
		"    j   markets          / new-york, chicago, topeka /;\n"+
		"\nParameters \n \n"+
		"    a(i)  capacity of plant i in cases\n    /   seattle     350\n        san-diego   600  /\n\n"+
		"    b(j)  demand at market j in cases\n    /   new-york    325\n        chicago     300\n        topeka      275  /;\n\n"+
		"Table d(i,j)  distance in thousands of miles \n"+
		"                new-york       chicago       topeka\n"+
		"    seattle          2.5           1.7          1.8\n"+
		"    san-diego        2.5           1.8          1.4 ;\n\n"+
		"Scalar f  freight in dollars per case per thousand miles  /90/;\n";
	return data;
}

static System::String^ GetModelText()
{
	System::String^ model ="Sets \n"+
		"    i   canning plants \n"+
		"    j   markets \n"+
		"\n"+
		"Parameters \n"+
		"    a(i)   capacity of plant i in cases \n"+
		"    b(j)   demand at market j in cases \n"+
		"    d(i,j) distance in thousands of miles \n"+
		"    Scalar f  freight in dollars per case per thousand miles; \n"+
		"\n"+
		"$if not set incname $abort 'no include file name for data file provided'\n"+
		"$include %incname% \n"+
		"\n"+
		"Parameter c(i,j)  transport cost in thousands of dollars per case ; \n"+
		"\n"+
		"    c(i,j) = f * d(i,j) / 1000 ; \n"+
		"\n"+
		"Variables \n"+
		"    x(i,j)  shipment quantities in cases \n"+
		"    z       total transportation costs in thousands of dollars ; \n"+
		"\n"+
		"Positive Variable x ; \n"+
		"\n"+
		"Equations \n"+
		"    cost        define objective function \n"+
		"    supply(i)   observe supply limit at plant i \n"+
		"    demand(j)   satisfy demand at market j ; \n"+
		"\n"+
		"cost ..        z  =e=  sum((i,j), c(i,j)*x(i,j)) ; \n"+
		"\n"+
		"supply(i) ..   sum(j, x(i,j))  =l=  a(i) ; \n"+
		"\n"+
		"demand(j) ..   sum(i, x(i,j))  =g=  b(j) ; \n"+
		"\n"+
		"Model transport /all/ ; \n"+
		"\n"+
		"Solve transport using lp minimizing z ; \n"+
		"\n"+
		"Display x.l, x.m ;\n";

	return model;
}



int main(array<String ^> ^args)
{
	GAMSWorkspace^ ws;
	if (args->Length > 0) 
		ws = gcnew GAMSWorkspace(nullptr,args[0],false);
	else
		ws = gcnew GAMSWorkspace(nullptr,nullptr,false);

	// write an include file containing data with GAMS syntax
	{
		StreamWriter^ writer = gcnew StreamWriter(ws->WorkingDirectory + Path::DirectorySeparatorChar + "tdata.gms");
		{
			writer->Write(GetDataText());
			writer->Close();
		}
	}
	// run a job using an instance of GAMSOptions that defines the data include file
	{
		GAMSOptions^ opt = ws->AddOptions(nullptr);
		{
			GAMSJob^ t2 = ws->AddJobFromString(GetModelText(),nullptr,nullptr);
			opt->Defines->Add("incname", "tdata");
			t2->Run(opt);
			for each (GAMSVariableRecord^ rec in t2->OutDB->GetVariable("x"))
				Console::WriteLine("x(" + rec->Keys[0] + "," + rec->Keys[1] + "): level=" + rec->Level + " marginal=" + rec->Marginal);
		}
	}
	return 0;
}

